File: C/1154_day_of_the_year.c
#include <string.h>

static const int days_passed[] = {0,31,59,90,120,151,181,212,243,273,304,334};

int dayOfYear(char* date) {
    int y, m, d, doy;
    if (!date || strlen(date) < 9)
        return -1;
    y = (date[0] - '0') * 1000 + (date[1] - '0') * 100 + 
	    (date[2] - '0') * 10 + 
	    (date[3] - '0');
    m = (date[5] - '0') * 10 + (date[6] - '0');
    d = (date[8] - '0') * 10 + (date[9] - '0');
    doy = days_passed[m - 1] + d;
    if (m > 2 && ((y % 4 == 0 && y % 100 != 0) || y % 400 == 0))
        doy++;
    return doy;
}

File: 1154_day_of_the_year.py
class Solution(object):
    def dayOfYear(self, date):
        """
        :type date: str
        :rtype: int
        """
        # days for previous months
        days_passed = [0,31,59,90,120,151,181,212,243,273,304,334]
        y, m, d = (int(d) for d in date.split('-'))
        doy = days_passed[m - 1] + d
        # A leap year must be divisble by 4 and not divisible by 100
        # or divisible by 400
        if m > 2 and ((y % 4 == 0 and y % 100 != 0) or y % 400 == 0):
            doy += 1
        return doy

vectors = [
        "2019-01-09", 9,
        "2019-02-10", 41,
        "1904-03-05", 65
        ]

for i in range(0, len(vectors), 2):
    date = vectors[i]
    doy = vectors[i + 1]
    print(f'{date} DOY is {doy}')
    returned = Solution().dayOfYear(date)
    assert returned == doy, f'for {date} day of the year is {doy}, returned {returned}!'
